text/0000-build-std/4-stage-1a.md

```diff
@@ -0,0 +1,1003 @@
+will not be used unless explicitly set and the crate graph's dependencies on the
+standard library will determine which crates are built instead. Otherwise,
+`build-std-crate` will default to "std".
+
+If `std` is to be built and Cargo is building a test using the default test
+harness then Cargo will also build the `test` crate.
+
+> [!NOTE]
+>
+> Inspired by the concept of [opaque dependencies][Opaque dependencies], the
+> standard library is resolved differently to other dependencies:
+>
+> - The lockfile included in the standard library source will be used when
+>   resolving the standard library's dependencies ([?][rationale-lockfile]).
+>
+> - The dependencies of the standard library crates are entirely opaque to the
+>   user. Different semver-compatible versions of these dependencies can
+>   exist in the user's resolve. The user cannot control compilation any of
+>   the dependencies of the `core`, `alloc` or `std` standard library crates
+>   individually (via profile overrides, for example).
+>
+> - The profile defined by the standard library will be used.
+>
+> Cargo will resolves the dependencies of opaque dependencies, such as the
+> standard library, separately in their own workspaces. The root of such a
+> resolve will be the crate specified in `build-std-crates`, or, if stage 1b is
+> implemented, the unified set of packages that any crate in the dependency has
+> a direct dependency on. A dependency on the relevant roots are added to all
+> crates in the "parent" resolve.
+>
+> Regardless of which standard library crates are being built, Cargo will build
+> the `sysroot` crate of the standard library workspace. `alloc` and `std` will
+> be optional dependencies of the `sysroot` crate which will be enabled when the
+> user has requested them. Panic runtimes are dependencies of `std` and will be
+> enabled depending on the features that Cargo passes to `std` (see
+> [*Panic strategies*][panic-strategies]).
+>
+> rustc loads panic runtimes in a different way to most dependencies, and
+> without looking in the sysroot they will fail to load correctly unless passed
+> in with `--extern`. rustc will need to be patched to be able to load panic
+> runtimes from `-L dependency=` paths in line with other transitive
+> dependencies.
+>
+> The standard library will always be a non-incremental build
+> ([?][rationale-incremental]), with no `depinfo` produced, and only a `rlib`
+> produced (no `dylib`) ([?][rationale-no-dylib]). It will be built in the Cargo
+> `target` directory of the crate or workspace like any other dependency.
+
+The host pre-built standard library will always be used for procedural macros
+and build scripts ([?][rationale-sysroot-for-host-deps]). Multi-target projects
+(resulting from the `target` field in artifact dependencies or the use of
+`per-pkg-target` fields) may result in the standard library being built multiple
+times - once for each target in the project.
```

Comment:
What about projects built without `--target`?  Those are being built in "host mode" rather than "cross-compiled" mode.  Should we be building everything for the host in the same way?

Reply:
At least for proc macros, it's important that we always use the host toolchain's prebuilt standard library so that the ABI matches that of the standard library rustc was linked with when the proc macro is loaded into the compiler. I haven't looked at the proc macro bridge in a long time but it is possible there are other considerations beyond just ABI compatibility that would require the standard libraries are exactly the same binary, not just ABI compatible.

Reply:
@adamgemmell that seems to be focused on "why can't build scripts / proc-macros do X" while instead my question was about "why don't other host mode builds have the same restriction?".

Reply:
I've noted the inconsistency with sharing host-mode dependencies in https://github.com/davidtwco/rfcs/pull/2/commits/488aafb306351cf7e7a09f244b3511415db7a99b

Reply:
Tweaked the wording for this slightly in 22a9b855136cd99004e60637f47659fedee32d2b

Reply:
Thanks!